File: 2022/19/java/Main.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class Main {
	// --- Constants ---
	private static final String INPUT_FILE = "../input/input.txt";
	private static final int PUZZLE1_TIME = 24;
	private static final int PUZZLE2_TIME = 32;
	private static final int PUZZLE2_BPS = 3;

	public static void main(String[] args) throws IOException, InterruptedException {
		List<Blueprint> bps = new ArrayList<>();

		// --- Reads and Parse the input file ---
		try (BufferedReader reader = new BufferedReader(new FileReader(INPUT_FILE))) {
			String line;
			while ((line = reader.readLine()) != null) {
				bps.add(new Blueprint(line));
			}
		}

		// --- Puzzle 1 ---
		SimThread[] threads = new SimThread[bps.size()];
		for (int i = 0; i < bps.size(); i++) {
			threads[i] = new SimThread(new Simulator(bps.get(i), PUZZLE1_TIME));
		}
		for (SimThread thread: threads) thread.start();
		for (SimThread thread: threads) thread.join();
		int total = 0;
		for (SimThread thread: threads) total += thread.getQualityLevel();
		System.out.format("1. Total sum of quality levels: %,d%n", total);

		// --- Puzzle 2 ---
		threads = new SimThread[PUZZLE2_BPS];
		for (int i = 0; i < threads.length; i++) {
			threads[i] = new SimThread(new Simulator(bps.get(i), PUZZLE2_TIME));
		}
		for (SimThread thread: threads) thread.start();
		for (SimThread thread: threads) thread.join();
		total = 1;
		for (SimThread thread: threads) total *= thread.getMax();
		System.out.format("2. Total of multiply the first 3 blurprints max values: %,d%n", total);
	}
}

File: 2022/19/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <file_reader.h>
#include <lib.h>

// --- Constants ---
#define INPUT_FILE "../input/input.txt"

// --- Macros ---
#define MAX(a, b, c, d) (a > b ? (a > c ? (a > d ? a : d) : c) : b)

// --- Main function ---
int main(void) {
	// --- Variables ---
	READ_BY_LINE_HEAD(file_t);
	READ_BY_LINE_INIT(file, file_t, INPUT_FILE)
	const int len = file.rows;
	Blueprint bps[len];
	int id, ore_ore, clay_ore, obs_ore, obs_clay, geo_ore, geo_obs;
	int i = 0, j = 0, k = 0;
	char line[file.max];

	// --- Read and parse the input file ---
	READ_BY_LINE_WHILE(file, line)
		sscanf(line, "Blueprint %d: Each ore robot costs %d ore. Each clay\
			robot costs %d ore. Each obsidian robot costs %d ore and %d clay.\
			Each geode robot costs %d ore and %d obsidian.", &id, &ore_ore,
			&clay_ore, &obs_ore, &obs_clay, &geo_ore, &geo_obs);

		bps[i].id = id;
		for (j = 0; j < 4; j++) for (k = 0; k < 3; k++) bps[i].costs[j][k] = 0;
		bps[i].costs[ORE][ORE] = ore_ore;
		bps[i].costs[CLAY][ORE] = clay_ore;
		bps[i].costs[OBSIDIAN][ORE] = obs_ore;
		bps[i].costs[OBSIDIAN][CLAY] = obs_clay;
		bps[i].costs[GEODE][ORE] = geo_ore;
		bps[i].costs[GEODE][OBSIDIAN] = geo_obs;
		bps[i].max_ore = MAX(ore_ore, clay_ore, obs_ore, geo_ore);
		i++;
	READ_BY_LINE_DONE(file)

	// --- Puzzle 1 ---
	int total = 0;
	for (i = 0; i < len; i++) total += bps[i].id * get_max(&bps[i], 24);
	printf("1. Total sum of quality levels: %d\n", total);

	total = 1;
	for (i = 0; i < 3; i++) total *= get_max(&bps[i], 32);
	printf("2. Total of multiply the first 3 blueprints max values: %d\n", total);

	return EXIT_SUCCESS;
}
